win32/win32.c

```diff
@@ -2615,9 +2615,73 @@ set_pioinfo_extra(void)
      * * https://bugs.ruby-lang.org/issues/18605
      */
     char *p = (char*)get_proc_address(UCRTBASE, "_isatty", NULL);
-    char *pend = p;
     /* _osfile(fh) & FDEV */
 
+#ifdef _M_ARM64
+#define IS_INSN(pc, name) ((*(pc) & name##_mask) == name##_id)
+    const int max_num_inst = 500;
+    uint32_t *start = (uint32_t*)p;
+    uint32_t *end_limit = (start + max_num_inst);
+    uint32_t *pc = start;
+
+    if (!p) {
+        fprintf(stderr, "_isatty proc not found in " UCRTBASE "\n");
+        _exit(1);
+    }
+
+    /* end of function */
+    const uint32_t ret_id = 0xd65f0000;
+    const uint32_t ret_mask = 0xfffffc1f;
+    for(; pc < end_limit; pc++) {
+        if (IS_INSN(pc, ret)) {
+            break;
+        }
+    }
+    if (pc == end_limit) {
+        fprintf(stderr, "end of _isatty not found in " UCRTBASE "\n");
+        _exit(1);
+    }
+
+    /* pioinfo instruction mark */
+    const uint32_t adrp_id = 0x90000000;
+    const uint32_t adrp_mask = 0x9f000000;
+    const uint32_t add_id = 0x11000000;
+    const uint32_t add_mask = 0x3fc00000;
```

Comment:
I'm having trouble understanding the add id/mask.  I think this is trying to match https://developer.arm.com/documentation/ddi0602/2024-09/Index-by-Encoding/Data-Processing----Immediate?lang=en#addsub_imm.  
```
0  0  0  1  0  0  0  1  0  0
0  0  1  1  1  1  1  1  1  1
sf op S  1  0  0  0  1  0  sh
```
So you mask off `sf` (size, 32 or 64 bit), and `op` (apparently add vs subtract), and match `S` must be off (not setting flags) and `sh` must be off (not shifting by 12 bits).  You then later implement checking and handling the `sh` bit left shifting by 12, but never check or implement the `op` bit turning it into a subtract.  Perhaps you meant mask `0x7f800000` instead?  Or am I off by one somewhere?

Reply:
I used doc pages for instructions encoding: [add](https://developer.arm.com/documentation/ddi0602/2022-06/Base-Instructions/ADD--immediate---Add--immediate--) - vs [sub](https://developer.arm.com/documentation/ddi0602/2022-06/Base-Instructions/SUB--immediate---Subtract--immediate--).

`_mask` are the bits that matters (same between sub and add).
`_id` are the bits expected, i.e. the concrete encoding. (bit 30 allow to differentiate sub vs add).

To identify an add, we need to check bits 23 to 30, which represent the instruction (fixed) encoding.
add_id: 0x11000000 (bits 23 to 30 are 0b00100010, sub_id would be 0b10100010)
add_mask: 0x3fc00000 (0b0111111110000000000000000000000, bits 23 to 30), sub_mask would be the same.

Note: The code included here is not supposed to be a complete disassembler, just a simple way to find a couple of sequences expected (it's supposed that only immediates may be changed in the future). So it makes assumptions (S bit at 0) and is "enough" to work, no more.

Reply:
> `_mask` are the bits that matters to identify encoding (same between sub and add). `_id` are the bits expected, i.e. the concrete encoding. (bit 30 allow to differentiate sub vs add).
> 
> To identify an add, we need to check bits 23 to 30, which represent the instruction (fixed) encoding. `add_id`: 0x11000000 (bits 23 to 30 are 0b00100010, `sub_id` would be 0b10100010) `add_mask`: 0x3fc00000 (0b0111111110000000000000000000000, bits 23 to 30), sub_mask would be the same.

`sub     x1, x2, #0x159` is:
`0xd1056441` is `0b11010001000001010110010001000001`
your mask is:
`0x3fc00000` is `0b00111111110000000000000000000000`
`and`ing them results in:
`11000000` is the same as `add_id`.

Reply:
oops, you're right, it seems that the mask is shifted to the right, and missing the `op` bit which is wrong.
And counting again yesterday, I did the same mistake...

The correct mask I get is `0x7f800000` (not including the sh bit). (0xff << 23, 23 being bits 23 to 30).

Reply:
Feel free to make a PR for it (and ping me, I can give you a +1).